lib/cfndsl/orchestration_template.rb

```diff
@@ -165,6 +186,132 @@ def check_output_refs
+        end
+
+        depends_on = resource.visit_json(path).with_object(Set.new(resource.depends_on)) do |(sub_path, value), resource_depends_on|
+          if value.nil?
+            errors << "Null value at #{sub_path} is not permitted"
+          elsif !(value_refs = value.respond_to?(:refs) ? value.refs - GLOBAL_REFS.keys : []).empty?
+            resource_refs = value_refs - parameters.keys
+            invalid_refs = resource_refs - resources.keys
+            if invalid_refs.empty?
+              resource_depends_on.merge(resource_refs) # Don't check for cyclic dependencies where references are not valid anyway
+            else
+              errors << "Invalid References : Path #{sub_path} refers to #{invalid_refs}"
+            end
+
+          end
+        end
+
+        resource_dependencies[logical_id] = depends_on
+      end
+
+      _validate_dependency_cycles(dependencies, errors)
+
+      errors
+    end
+    # rubocop:enable  Metrics/AbcSize, Metrics/PerceivedComplexity
+
+    # @api private
+    # rubocop:disable Metrics/PerceivedComplexity
+    def _validate_dependency_cycles(dependencies, errors = [], logical_id = nil, visited = Set.new)
+      if logical_id
+        visited << logical_id
+        depends_on = dependencies[logical_id]
+        if (cycles = depends_on & visited).any?
+          errors << "Found cyclic dependency for #{logical_id} to #{cycles.to_a}"
+        else
+          depends_on.each do |depends_on_logical_id|
+            if dependencies.key?(depends_on_logical_id)
+              _validate_dependency_cycles(dependencies, errors, depends_on_logical_id, visited)
+            else
+              errors << "#{logical_id} DependsOn unknown resource #{depends_on_logical_id}"
+            end
+            break unless errors.empty?
+          end
+        end
+      else
+        dependencies.keys.each do |dependency_logical_id|
+          _validate_dependency_cycles(dependencies, errors, dependency_logical_id)
+          break unless errors.empty?
+        end
+      end
+    end
+    # rubocop:enable Metrics/PerceivedComplexity
+
+    # rubocop:disable Metrics/PerceivedComplexity
+    def validate_conditions(errors = [])
+      conditions = (@Conditions || {})
+      parameter_refs = (@Parameters || {}).keys
+
+      conditions.each_pair do |logical_id, condition|
+        path = "/Conditions/#{logical_id}"
+
+        condition.visit_json(path) do |sub_path, value|
+          if value.nil?
+            errors << "Null value at #{sub_path} is not permitted"
+          elsif !(value_refs = value.respond_to?(:refs) ? value.refs - GLOBAL_REFS.keys : []).empty?
+            invalid_refs = value_refs - parameter_refs
+            errors << "Invalid References : Path #{sub_path} refers to #{invalid_refs}" unless invalid_refs.empty?
+          end
+        end
+      end
+
+      # TODO: condition functions Fn::Not, Fn::Or, Fn::And can refer to other conditions, and these cannot be cyclic either
```

Comment:
What is the policy on TODOs ? Implementing this check would require more complexity in CfnDsl::Fn - likely separating condition functions from others.